### PR TITLE
Ltd 3563 optimise cases search

### PR DESCRIPTION
**Pull request by MarkThomasLite**
### Aim

Reduce queries and duration when loading cases search (1000 DB queries and ~6.5s duration on local).

work was focused on the following:
- populate_destinations
- populate_activity_updates

queries down to 125 and ~1.1s duration, again on local

[LTD-3563](https://uktrade.atlassian.net/browse/LTD-3563)

[LTD-3563]: https://uktrade.atlassian.net/browse/LTD-3563?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: api/cases/views/search/service.py [hunk 1]
```diff
@@ -215,7 +215,8 @@ def populate_activity_updates(cases: List[Dict]):
     retrieve the last 2 activities per case for the provided list of cases
     """
     obj_type = ContentType.objects.get_for_model(Case)
-    case_ids = [case["id"] for case in cases]
+    case_map = {case["id"]: case for case in cases}
+    case_ids = list(case_map.keys())
```
Comment (kevincarrogan) [inline api/cases/views/search/service.py:222]: You appear to be doing this in both this function and the `populate_destinations`. I wonder if we could produce this map and the ids list just once outside this function and pass it in to both of them instead.

Comment (kevincarrogan) [reply api/cases/views/search/service.py:222]: You could also have a single loop that just iterates through the cases and produces both the keys list and the dictionary for even MOAR PERFORMANCE GAInS.

Comment (MarkThomasLite) [reply api/cases/views/search/service.py:214]: will do, was trying to keep them all seperate as the next step would be a rewrite and reuse of objects between several of the populate methods

File: api/cases/views/search/service.py [hunk 2]
```diff
@@ -195,28 +187,26 @@ def get_hmrc_sla_hours(cases: List[Dict]):
...
+        application__in=list(case_map.keys()), deleted_at=None
+    )
+    for poa in poas:
+        serializer = CountrySerializer(poa.party.country)
+        data = serializer.data
+        case = case_map[str(poa.application_id)]
+        if case.get("destinations"):
+            case["destinations"].append({"country": data})
+        else:
+            case["destinations"] = [{"country": data}]
```
Comment (kevincarrogan) [inline api/cases/views/search/service.py:201]: I think you've instantiated this elsewhere now.

**Review by kevincarrogan — APPROVED** [review]

File: api/cases/views/search/service.py (hunk 2 above)
Comment (MarkThomasLite) [reply api/cases/views/search/service.py]: course as part of a subsequent bug fix

**Review by saruniitr — APPROVED** [review]

**Pull request merged**